MAINT: bump OpenBLAS version, bump macOS image used in GHA

andyfaff opened the pull request:
Bumps OpenBLAS version from 0.3.18 to 0.3.21. The changeset is a bit larger that may be expected because the bump comes with alterations to compiler paths, openblas paths, etc.

Note, a few test fails are expected. These will probably need to result in test tweaks to tolerances, etc. I don't believe that they're materially caused by this PR, just the blas bump itself.

[EDIT] - More context:
This PR bumps the macOS image used on Github Actions, both in wheel building and in general CI. Previously the `macOS-10.15` image was being used, it's now bumped to at least `macOS-11`. This is because the 10.15 image has been [unsupported](https://github.blog/changelog/2022-07-20-github-actions-the-macos-10-15-actions-runner-image-is-being-deprecated-and-will-be-removed-by-8-30-22/) for a long time .
I was having challenges using the older gfortran compiler that was being used in various places (gfortran-4.9) when bumping the image, so needed  to update the gfortran compiler being used. This gfortran bump also required a more recent OpenBLAS build than one from 2 years ago. I settled on 0.3.21 as this is what numpy is using. Bumping the OpenBLAS version has required some adjustments to test tolerances.

andyfaff commented:
supercedes #17370

rgommers commented:
The macOS setup.py failure shows:
```
E   ImportError: dlopen(/Users/runner/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/scipy-1.11.0.dev0+0.ad409b0-py3.8-macosx-10.15-x86_64.egg/scipy/sparse/linalg/_isolve/_iterative.cpython-38-darwin.so, 0x0002): Library not loaded: '@rpath/libgfortran.5.dylib'
E     Referenced from: '/usr/local/lib/libopenblas.0.dylib'
E     Reason: tried: '/usr/local/lib/libgfortran.5.dylib' (no such file), '/usr/local/lib/libgfortran.5.dylib' (no such file), '/usr/local/lib/libgfortran.5.dylib' (no such file), '/usr/lib/libgfortran.5.dylib' (no such file)
```
A little unclear what is going on there - is there a mismatch between the `gfortran` version with which the new `openblas` was built and the `gfortran` installed here? Either way, you may want to bump that job to Python 3.9, because we should be dropping 3.8 before the next release.

andyfaff commented:
The openblas version was built with a more recent version of gfortran, so the gfortran tool chain needs to be updated on macos.yml.
I think other CI entries may need to be updated as well, but I'll do that ina separate PR.

andyfaff commented:
I think I've solved the macos.yml entry. It needed an update to openblas/gfortran toolchain. It then couldn't find the gfortran library so I applied a hammer approach:
```
        export SDKROOT=/Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk
        export PATH=/usr/local/gfortran/lib:/opt/gfortran-darwin-x86_64-native/lib/:$PATH
        export LIBRARY_PATH=/usr/local/gfortran/lib:/opt/gfortran-darwin-x86_64-native/lib/:$LIBRARY_PATH
        export LD_LIBRARY_PATH=/usr/local/gfortran/lib:/opt/gfortran-darwin-x86_64-native/lib/:$LD_LIBRARY_PATH
        sudo cp /usr/local/gfortran/lib/*.dylib /usr/local/lib
```

I can try and figure out which one of the above fixed the issue if necessary (but would require a few more iterations).

andyfaff commented:
BTW, what's the current sunset period for setup.py? When can we remove all the setup.py entries from CI?

andyfaff commented:
I reported the lobpcg errors in #17954. There's also a further test failure on Linux 32 bit. I'm not familar with linalg, but it seems like a minor test tweak is also required here. If someone lets me know what to tweak, or wants to push to this PR, please do so.
```
2023-02-10T09:05:13.4466521Z _____________________ test_solve_generalized_discrete_are ______________________
2023-02-10T09:05:13.4467903Z [gw1] linux -- Python 3.8.0 /usr/bin/python3.8
2023-02-10T09:05:13.4469181Z /usr/local/lib/python3.8/dist-packages/scipy-1.11.0.dev0+unknown.unknown-py3.8-linux-i686.egg/scipy/linalg/tests/test_solvers.py:654: in test_solve_generalized_discrete_are
2023-02-10T09:05:13.4470110Z     _test_factory(case, min_decimal[ind])
2023-02-10T09:05:13.4470796Z         _test_factory = <function test_solve_generalized_discrete_are.<locals>._test_factory at 0xe14c27c0>
2023-02-10T09:05:13.4472282Z         case       = (array([[ 0.97905961,  0.15428665, -0.03994193,  0.00391164, -0.01180311,
2023-02-10T09:05:13.4473536Z         -0.02494093, -0.00892079,  0.0143359...891e-06,  2.94712339e-08],
2023-02-10T09:05:13.4474754Z        [ 2.40952333e-08,  4.30614350e-07],
2023-02-10T09:05:13.4475589Z        [ 1.23662598e-08, -6.06921957e-07]]), ...)
2023-02-10T09:05:13.4476227Z         cases      = [(array([[0.276923  , 0.8234578 , 0.950222  ],
2023-02-10T09:05:13.4476791Z        [0.04617139, 0.6948286 , 0.03444608],
2023-02-10T09:05:13.4477590Z        [0.09713178, 0.3170...91e-06,  2.94712339e-08],
2023-02-10T09:05:13.4478546Z        [ 2.40952333e-08,  4.30614350e-07],
2023-02-10T09:05:13.4479518Z        [ 1.23662598e-08, -6.06921957e-07]]), ...)]
2023-02-10T09:05:13.4480391Z         ind        = 2
2023-02-10T09:05:13.4481276Z         mat20170120 = {'A': array([[ 0.97905961,  0.15428665, -0.03994193,  0.00391164, -0.01180311,
2023-02-10T09:05:13.4482395Z         -0.02494093, -0.00892079,  0.01...,  2.58702432e-06]]), 'R': array([[ 6.86199839e-07, -1.42279806e-08],
2023-02-10T09:05:13.4483331Z        [-1.42279806e-08,  1.81130866e-07]]), ...}
2023-02-10T09:05:13.4483921Z         min_decimal = (11, 11, 16)
2023-02-10T09:05:13.4484977Z /usr/local/lib/python3.8/dist-packages/scipy-1.11.0.dev0+unknown.unknown-py3.8-linux-i686.egg/scipy/linalg/tests/test_solvers.py:651: in _test_factory
2023-02-10T09:05:13.4485928Z     assert_array_almost_equal(res, np.zeros_like(res), decimal=dec)
2023-02-10T09:05:13.4486497Z E   AssertionError: 
2023-02-10T09:05:13.4486993Z E   Arrays are not almost equal to 16 decimals
2023-02-10T09:05:13.4487470Z E   
2023-02-10T09:05:13.4487916Z E   Mismatched elements: 2 / 64 (3.12%)
2023-02-10T09:05:13.4488677Z E   Max absolute difference: 2.31270488e-16
2023-02-10T09:05:13.4489246Z E   Max relative difference: inf
2023-02-10T09:05:13.4490073Z E    x: array([[-6.7178183046738560e-18, -2.4140438996747560e-18,
2023-02-10T09:05:13.4490968Z E           -1.0194888553152759e-17, -5.0376755142357479e-17,
2023-02-10T09:05:13.4491872Z E           -5.0720332881587504e-17,  2.3244331078111713e-17,...
2023-02-10T09:05:13.4492524Z E    y: array([[0., 0., 0., 0., 0., 0., 0., 0.],
2023-02-10T09:05:13.4493046Z E          [0., 0., 0., 0., 0., 0., 0., 0.],
2023-02-10T09:05:13.4493573Z E          [0., 0., 0., 0., 0., 0., 0., 0.],...
2023-02-10T09:05:13.4494434Z         a          = array([[ 0.97905961,  0.15428665, -0.03994193,  0.00391164, -0.01180311,
2023-02-10T09:05:13.4495559Z         -0.02494093, -0.00892079,  0.01433595...     [-0.02851398, -0.00156313, -0.01257725, -0.05334177, -0.03769227,
2023-02-10T09:05:13.4496514Z         -0.00788838, -0.62801576,  0.34907   ]])
2023-02-10T09:05:13.4497322Z         b          = array([[-1.04368719,  1.09316659],
2023-02-10T09:05:13.4498110Z        [ 0.51336209, -0.15143908],
2023-02-10T09:05:13.4498650Z        [ 0.78618282,  0.3888125 ],
2023-02-10T09:05:13.4499137Z        [ 0.90... 0.12416864],
2023-02-10T09:05:13.4499810Z        [ 0.90485159, -0.0090001 ],
2023-02-10T09:05:13.4500572Z        [-0.00863767,  0.55645316],
2023-02-10T09:05:13.4501104Z        [ 0.00708354,  0.6553915 ]])
2023-02-10T09:05:13.4501982Z         case       = (array([[ 0.97905961,  0.15428665, -0.03994193,  0.00391164, -0.01180311,
2023-02-10T09:05:13.4503222Z         -0.02494093, -0.00892079,  0.0143359...891e-06,  2.94712339e-08],
2023-02-10T09:05:13.4504083Z        [ 2.40952333e-08,  4.30614350e-07],
2023-02-10T09:05:13.4504904Z        [ 1.23662598e-08, -6.06921957e-07]]), ...)
2023-02-10T09:05:13.4505441Z         dec        = 16
2023-02-10T09:05:13.4505950Z         e          = array([[1., 0., 0., 0., 0., 0., 0., 0.],
2023-02-10T09:05:13.4506460Z        [0., 1., 0., 0., 0., 0., 0., 0.],
2023-02-10T09:05:13.4507016Z        [0., 0., 1., 0., 0., 0., 0., ...   [0., 0., 0., 0., 0., 1., 0., 0.],
2023-02-10T09:05:13.4507570Z        [0., 0., 0., 0., 0., 0., 1., 0.],
2023-02-10T09:05:13.4508056Z        [0., 0., 0., 0., 0., 0., 0., 1.]])
2023-02-10T09:05:13.4508545Z         knownfailure = None
2023-02-10T09:05:13.4509342Z         q          = array([[ 1.75512507e-06, -4.27225221e-07, -9.34790486e-07,
2023-02-10T09:05:13.4510454Z         -6.10641258e-07,  1.62839572e-06,  2.22620136e-06,
2023-02-10T09:05:13.4511230Z ...-7.58710361e-07,
2023-02-10T09:05:13.4512249Z         -9.08221173e-07, -2.90720260e-07, -6.17890194e-08,
2023-02-10T09:05:13.4513150Z         -2.00190655e-06,  2.58702432e-06]])
2023-02-10T09:05:13.4513982Z         r          = array([[ 6.86199839e-07, -1.42279806e-08],
2023-02-10T09:05:13.4514806Z        [-1.42279806e-08,  1.81130866e-07]])
2023-02-10T09:05:13.4515680Z         res        = array([[-6.71781830e-18, -2.41404390e-18, -1.01948886e-17,
2023-02-10T09:05:13.4516598Z         -5.03767551e-17, -5.07203329e-17,  2.32443311e-17,
2023-02-10T09:05:13.4518207Z ...-1.54652070e-17,
2023-02-10T09:05:13.4519231Z         -3.89479249e-17,  7.98257084e-18, -7.57401310e-19,
2023-02-10T09:05:13.4520088Z          8.60363128e-17,  3.04918097e-17]])
2023-02-10T09:05:13.4520923Z         s          = array([[-6.27256153e-07,  2.80554657e-07],
2023-02-10T09:05:13.4521763Z        [ 4.27827567e-07,  8.54401686e-08],
2023-02-10T09:05:13.4522682Z        [ 8.76221935e-07,  1.5285....68302891e-06,  2.94712339e-08],
2023-02-10T09:05:13.4523549Z        [ 2.40952333e-08,  4.30614350e-07],
2023-02-10T09:05:13.4524337Z        [ 1.23662598e-08, -6.06921957e-07]])
2023-02-10T09:05:13.4525237Z         x          = array([[ 1.88357055e-06, -3.98902725e-07, -8.97987558e-07,
2023-02-10T09:05:13.4526154Z         -7.18984081e-07,  1.95043873e-06,  2.15596656e-06,
2023-02-10T09:05:13.4526908Z ...-7.62588561e-07,
2023-02-10T09:05:13.4527734Z         -8.85735630e-07, -2.59621085e-07, -7.62321103e-08,
2023-02-10T09:05:13.4528553Z         -1.79447605e-06,  3.08567024e-06]])
```

mattip commented:
There is also a failure on the linux wheel builds, an [expected warning is not emitted](https://github.com/scipy/scipy/actions/runs/4140432079/jobs/7159925764#step:5:5646)
```
../venv/lib/python3.8/site-packages/scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py:268: \
    in test_failure_to_run_iterations
      eigenvalues, _ = lobpcg(A, Q, maxiter=40, tol=1e-12)
  E   Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
  E   The list of emitted warnings is: [].
```

SeanChen1992 commented:
您的邮件我已收到，我会加班加点的看完您珍贵的来信，尽快给您回复的！谢谢啦！陈鑫

andyfaff commented:
@mattip, that should be solved by #17975

andyfaff commented:
Thinking of changing the linalg tolerance to `np.assert_allclose(..., atol=3.5e-16)`, is that reasonable?

rgommers commented:
@andyfaff that seems fine to me, that's still very tight.

rgommers commented:
> BTW, what's the current sunset period for setup.py? When can we remove all the setup.py entries from CI?

Hopefully well before the next release split. It requires me sitting down and finishing the BLAS/LAPACK detection features in Meson before the next Meson release.

rgommers commented:
@andyfaff it may be useful to summarize the macOS image/version related changes in the PR description. If I understand correctly: we must move to `macos-11` because the older image is getting dropped; the use of the `10.15` SDK is needed to get wheels that still work on macOS 10.15. What is less clear to me is whether or not that then also works on `<10.15` and what the final wheel tag should be.

isuruf commented on tools/wheels/cibw_before_build_macos.sh:
```diff
@@ -21,17 +21,26 @@ if [[ $PLATFORM == "macosx-x86_64" ]]; then
+  sudo cp -R /opt/gfortran-darwin-x86_64-native/lib/ /usr/local/gfortran/lib
+  ln -sf /opt/gfortran-darwin-x86_64-native/bin/gfortran /usr/local/bin/gfortran
+  export PATH=/usr/local/gfortran/lib:$PATH
+  export SDKROOT=/Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk
```
See https://github.com/MacPython/gfortran-install/blob/47bb0df1e6f766503f81e4fd3d5d2acadd887393/gfortran_utils.sh#L93-L94

andyfaff replied:
The GHA images have quite a few SDKs installed. Wouldn't the suggestion select the most recent SDK? I was trying to use the oldest SDK so that would keep the best compatibility with older versions of macOS (old MACOSX_DEPLOYMENT_TARGET).

isuruf replied:
Version of the SDK and the MACOSX_DEPLOYMENT_TARGET are two different things. You can use the latest SDK (13.1) and set MACOSX_DEPLOYMENT_TARGET=10.9 and it should work just fine.

isuruf commented on tools/wheels/cibw_before_build_macos.sh:
```diff
@@ -21,17 +21,26 @@ if [[ $PLATFORM == "macosx-x86_64" ]]; then
+  # bin, lib, include, libexec underneath that.
+  sudo tar -xv -C /opt -f gfortran.tar.gz
+  sudo mkdir -p /usr/local/gfortran/lib
+  sudo cp -R /opt/gfortran-darwin-x86_64-native/lib/ /usr/local/gfortran/lib
+  ln -sf /opt/gfortran-darwin-x86_64-native/bin/gfortran /usr/local/bin/gfortran
```
You probably want to do https://github.com/MacPython/gfortran-install/blob/47bb0df1e6f766503f81e4fd3d5d2acadd887393/gfortran_utils.sh#L125-L130 instead

andyfaff replied:
Can do. However, the overall effect seems quite similar? What are the advantages of the suggestion? (coming at this from a learning POV)

isuruf replied:
`/usr/local/lib` is in the linker and the dynamic loader's search path by default. So you probably don't have to mess with `LIBRARY_PATH` and `DYLD_LIBRARY_PATH`.

andyfaff commented:
This PR bumps the macOS image used on Github Actions, both in wheel building and in general CI. Previously the `macOS-10.15` image was being used, it's now bumped to at least `macOS-11`. This is because the 10.15 image has been [unsupported](https://github.blog/changelog/2022-07-20-github-actions-the-macos-10-15-actions-runner-image-is-being-deprecated-and-will-be-removed-by-8-30-22/) for a long time .
I was having challenges using the older gfortran compiler that was being used in various places (gfortran-4.9) when bumping the image, so needed  to update the gfortran compiler being used. This gfortran bump then required a more recent OpenBLAS build than one from 2 years ago. I settled on 0.3.21 as this is what numpy is using. Bumping the OpenBLAS version has required some adjustments to test tolerances.

@rgommers, the `macosx_x86_64` wheels being uploaded to nightly were tagged as `10_9` up until two months ago. The current nightly wheels are tagged `10_15`. I tried looking through changes in various infrastructure places, but I don't understand why that change happened. As I understand it the compiler chain should still support 10.9, so will try and get the wheels tagged as 10.9 again.

isuruf commented on tools/wheels/cibw_before_build_macos.sh:
```diff
@@ -34,12 +34,15 @@ if [[ $PLATFORM == "macosx-x86_64" ]]; then
+
+  # Link these into /usr/local so that there's no need to add rpath or -L
+  for f in libgfortran.dylib libgfortran.5.dylib libgcc_s.1.dylib libgcc_s.1.1.dylib libquadmath.dylib libquadmath.0.dylib; do
+    ln -sf /opt/gfortran-darwin-${arch}-${type}/lib/$f /usr/local/lib/$f
```
```suggestion
    ln -sf /opt/gfortran-darwin-x86_64-native/lib/$f /usr/local/lib/$f
```

andyfaff replied:
thanks for that catch

andyfaff commented:
rebased to pull in the lobpcg test fix.
Added the `MACOSX_DEPLOYMENT_TARGET=10.9` and `_PYTHON_HOST_PLATFORM="macosx-12.0-arm64"` environment variables to try and enforce the 10_9 tag in the `macosx_x86_64` entries. I don't have access to macos<10.15 to test though.

andyfaff commented:
Still not creating the correct tag for macos, even with `MACOSX_DEPLOYMENT_TARGET` set to 10.9. This might be related to https://github.com/pypa/cibuildwheel/issues/1419

andyfaff commented:
Specifying `MACOS_DEPLOYMENT_TARGET` worked, the macOS wheels are now tagged correctly. The issue was in meson-python and is fixed in this [merged PR](https://github.com/mesonbuild/meson-python/pull/309). Once there's a new release the `MACOS_DEPLOYMENT_TARGET` variable in wheels.yml can be removed.

andyfaff commented:
@rgommers, I believe we can still use the `10_9` tag. Wheels tagged as `10_9` should've been produced ever since we started using cibuildwheel. That there are `10_15` tags is because of the previously mentioned bug in `meson-python`.

When wheels are next uploaded to nightly can you remove those wheels tagged as `10_15`, they'll be stale. We wouldn't want them being picked up over more recent `10_9` files. 

The meson-python bug is being experienced in maintenance branches as well. If I look at [multibuild-wheels-staging](https://anaconda.org/multibuild-wheels-staging/scipy/files) there are wheels tagged as `10_15`. It may be worth making a one-liner PR against the 1.10 maintenance branch to fix the issue in the next bug release. It may be necessary to remove the wrong tags from staging as well.
 @tylerjereddy

h-vetinari commented on tools/openblas_support.py:
```diff
@@ -54,39 +55,67 @@ def get_ilp64():
     # XXX For PEP 600 this can be a glibc version
-    assert ret in ('1', '2010', '2014', '_2_24'), f'invalid MB_ML_VER {ret}'
-    return ret
+    assert ml_ver in ('2010', '2014', '_2_24'), f'invalid MB_ML_VER {ml_ver}'
+    suffix = f'manylinux{ml_ver}_{arch}.tar.gz'
+    return suffix
```
This might be a question for multibuild first, but we should eventually add `_2_28` here, as `_2_24` is EOL and has effectively been replaced by `_2_28`.

andyfaff replied:
At the moment the only `ml_ver` options on https://anaconda.org/multibuild-wheels-staging/openblas-libs/files are `{'2010', '2014'}`

mattip replied:
Multibuild supports `_2_28`. As for openblas-libs, I could build `_2_28` tarballs, but as 2014 will only go EOL in June 2024, I would prefer not to right now. As far as I know, there are no glibc dependencies in OpenBLAS, so there is no compelling reason to update.

I think I started building a `_2_24` openblas-libs tarball at one point but reached the conclusion that it was not needed, so line 64 could be limited to the versions currently on anaconda.org: 2010 and 2014 (and maybe even just 2014 alone, I may drop 2010 soon).

rgommers commented on tools/wheels/cibw_before_build_macos.sh:
```diff
@@ -21,17 +21,31 @@ if [[ $PLATFORM == "macosx-x86_64" ]]; then
+
+  # Set SDKROOT env variable if not set
+  # This step is required whenever the gfortran compilers sourced from
+  # conda-forge (built by isuru fernando) are used
```
This was a little confusing, so I propose adding: "outside of a conda-forge environment (so it mirrors what is done in the conda-forge compiler activation scripts)"

andyfaff replied:
The compiler tarball that @isuruf makes are collections of individual files extracted from conda-forge builds. They require the SDKROOT variable to be set otherwise they don't work. I'll rework the comment.

rgommers commented on .github/workflows/macos.yml:
```diff
@@ -89,7 +61,9 @@ jobs:
     - name: Test SciPy
       run: |
-        export LIBRARY_PATH="$LIBRARY_PATH:/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib"
+        # Set SDKROOT env variable if not set
```
Can you refer to the more extensive comment in the wheel build yml file here?

rgommers reviewed (commented):
This is looking pretty good, thanks @andyfaff. A few minor comments.

andyfaff commented on azure-pipelines.yml:
```diff
@@ -274,19 +274,19 @@ stages:
+
+    - powershell: |
+        # wheels must use same version
+        choco install -y mingw --forcex86 --force --version=8.1.0
```
I don't think that the setup.py build actually used clang-cl as the c-compiler. If I look at an azure build log, e.g.

https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=23257&view=logs&j=203f29d8-a99b-512b-bad4-9c985ff1792a&t=549b98f4-8390-5ed6-827a-12915e592e34&l=9317

the existing builds appear to use MSVC for the compilation. Or is that an incorrect take?

andyfaff replied:
I think that choco installing mingw probably might not required for the 64-bit build, the (gfortran, gcc) compilers are actually installed via mingw in rtools. I know that those tools worked for the wheels so I used them here.

I used the 8.1.0 for the 32 bit install because the OpenBLAS 32-bit library was built with gcc 8.3.0 (the closest version). 8.1.0 is quite an old gcc version, so it made less sense to try and get the 64-bit build to work with 8.1.0.

I had to do lots of debugging in CI of the azure builds because they're quite old setups, so once I got something working I left it alone.

rgommers replied:
`clang-cl.exe` is used by Pythran, see for example https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=23257&view=logs&j=d30ccee0-62b2-572b-86e1-0fd5d4704592&t=dd6887a0-9c47-504f-8c01-87a519f6f92e. You linked to the job that disabled Pythran.

> I had to do lots of debugging in CI of the azure builds because they're quite old setups, so once I got something working I left it alone.

Okay, that sounds fine for the 32-bit builds.

andyfaff replied:
> clang-cl.exe is used by Pythran

So why doesn't Pythran complain if clang-cl isn't installed? Can it not use the C compiler from the rest of the toolchain? Is the solution here just to install clang-cl again?

rgommers replied:
> So why doesn't Pythran complain if clang-cl isn't installed? Can it not use the C compiler from the rest of the toolchain? Is the solution here just to install clang-cl again?

It can use Mingw just fine, we should only make sure that it does not use MSVC because that leads to hard-to-debug problems.

andyfaff replied:
`setup.py == PITA`. I tried a multitude of ways when creating this PR to avoid MSVC and force the use of gcc. `setup.py` seems to ignore the environment variables `CC` and `CXX` on Windows. Anyway abe9627 seems to be using clang-cl for Pythran compilation.

rgommers replied:
Meson is kind of the opposite - it's good with `CC` & co, but you have to work a bit to make it use MSVC because it's not on the PATH: https://dvdhrm.github.io/2021/04/21/meson-msvc-github-actions/.

rgommers commented:
Funky logging error happening https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=23260&view=logs&jobId=203f29d8-a99b-512b-bad4-9c985ff1792a&j=203f29d8-a99b-512b-bad4-9c985ff1792a&t=0e9bc6fa-f840-5c44-330b-bc382f4fd289:
```
--- Logging error ---
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\logging\__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: underlying buffer has been detached
Call stack:
  File "setup.py", line 533, in <module>
    setup_package()
  File "setup.py", line 529, in setup_package
    setup(**metadata)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\numpy\distutils\core.py", line 169, in setup
    return old_setup(**new_attr)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\setuptools\__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\wheel\bdist_wheel.py", line 387, in run
    wf.write_files(archive_root)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\zipfile.py", line 1312, in __exit__
    self.close()
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\wheel\wheelfile.py", line 189, in close
    self.writestr(zinfo, data.getvalue())
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\site-packages\wheel\wheelfile.py", line 165, in writestr
    log.info(f"adding '{fname}'")
Message: "adding 'scipy-1.11.0.dev0+1540.e3925ef.dist-info/RECORD'"
Arguments: ()
```

Not sure what's up with that, seems like a `bdist_wheel` related issue. All CI still passes though. Since I see that issue on other PRs as well, I'll ignore it here.

rgommers reviewed (approved):
CI is all green and this looks like it's in good shape. Very nice upgrade, thanks a lot @andyfaff!

isuruf reviewed (commented):
Btw, you can use the `arm64-native` toolchain for arm64 so that you only have to maintain build scripts for only one toolchain instead of two. (`isuruf/gcc` and `fxcoudert/gfortran-for-macOS`)
